core/lib/src/request/from_request.rs

```diff
@@ -207,6 +207,118 @@ impl<S, E> IntoOutcome<S, (Status, E), ()> for Result<S, E> {
+/// #     type Error = ();
+/// #     fn from_request(request: &'a Request<'r>) -> request::Outcome<Database, ()> {
+/// #         Outcome::Success(Database)
+/// #     }
+/// # }
+/// #
+/// # struct Admin { user: User };
+/// #
+/// impl<'a, 'r> FromRequest<'a, 'r> for User {
+///     type Error = ();
+///
+///     fn from_request(request: &'a Request<'r>) -> request::Outcome<User, ()> {
+///         let db = request.guard::<Database>()?;
+///         request.cookies()
+///             .get_private("user_id")
+///             .and_then(|cookie| cookie.value().parse().ok())
+///             .and_then(|id| db.get_user(id).ok())
+///             .or_forward(())
+///     }
+/// }
+///
+/// impl<'a, 'r> FromRequest<'a, 'r> for Admin {
+///     type Error = ();
+///
+///     fn from_request(request: &'a Request<'r>) -> request::Outcome<Admin, ()> {
+///         // This will unconditionally query the database!
+///         let user = request.guard::<User>()?;
+///
+///         if user.is_admin {
+///             Outcome::Success(Admin { user })
+///         } else {
+///             Outcome::Forward(())
+///         }
+///     }
+/// }
+///
+/// #[get("/dashboard")]
+/// fn admin_dashboard(admin: Admin) { }
+///
+/// #[get("/dashboard", rank = 2)]
+/// fn user_dashboard(user: User) { }
+/// ```
+///
+/// When a non-admin user is logged in, the database will be queried twice: Once
+/// via the `Admin` guard invoking the `User` guard, and a second time via the
+/// `User` guard directly. For cases such as these, the request-local cache
+/// should be used:
+///
+/// ```
+/// # #![feature(plugin, decl_macro)]
+/// # #![plugin(rocket_codegen)]
+/// # extern crate rocket;
+/// #
+/// # use rocket::outcome::{IntoOutcome, Outcome};
+/// # use rocket::request::{self, FromRequest, Request};
+/// # struct User { id: String, is_admin: bool };
+/// # struct Database;
+/// # impl Database {
+/// #     fn get_user(&self, id: String) -> Result<User, ()> {
+/// #         Ok(User { id, is_admin: false })
+/// #     }
+/// # }
+/// # impl<'a, 'r> FromRequest<'a, 'r> for Database {
+/// #     type Error = ();
+/// #     fn from_request(request: &'a Request<'r>) -> request::Outcome<Database, ()> {
+/// #         Outcome::Success(Database)
+/// #     }
+/// # }
+/// #
+///
+/// impl<'a, 'r> FromRequest<'a, 'r> for &'a User {
+///     type Error = ();
+///
+///     fn from_request(request: &'a Request<'r>) -> request::Outcome<&'a User, ()> {
+///         request.local_cache(|| {
+///             // This closure will run only once per request, and future
+///             // invocations will reuse the result of the first calculation
+///             let db = request.guard::<Database>().succeeded()?;
+///             request.cookies()
+///                 .get_private("user_id")
+///                 .and_then(|cookie| cookie.value().parse().ok())
+///                 .and_then(|id| db.get_user(id).ok())
+///         }).as_ref().or_forward(())
```

Comment:
The return type and last line of this function aren't very obvious but are necessary given how `local_cache` works. It also means that `impl<'a, 'r> FromRequest<'a, 'r> for Admin` must be changed to something with a lifetime, such as `for Admin<'a>`.

Reply:
Yeah, that's pretty interesting.

I would put the comment outside of the call (for style/readability purposes). And then, I _might_ do:

```rust
let user_result = request.local_cache(|| {
    // everything
});

user_result.as_ref().or_forward(())
```

I'd also add a postlude to this example with information on what's going on with the references here, which is pretty interesting. Also, it would help to complete the example by including the `Admin` guard as well.